src/header/common/content_disposition.rs

```diff
@@ -65,88 +218,215 @@ pub enum DispositionParam {
+                let param_name = &param_name[..param_name.len() - 1]; // trim asterisk
+                let (ext_value, new_left) = split_once_and_trim(left, ';');
+                left = new_left;
+                let ext_value = header::parse_extended_value(ext_value)?;
+                cd.parameters
+                    .push(match param_name.to_ascii_lowercase().as_str() {
+                        "filename" => DispositionParam::FilenameExt(ext_value),
+                        _ => DispositionParam::UnknownExt(
+                            param_name.to_owned(),
+                            ext_value,
+                        ),
+                    });
+            } else {
+                // regular parameters
+                let value = if left.starts_with('\"') {
+                    // quoted-string: defined in RFC6266 -> RFC2616 Section 3.6
+                    let mut escaping = false;
+                    let mut quoted_string = vec![];
+                    let mut end = None;
+                    // search for closing quote
+                    for (i, &c) in left.as_bytes().iter().skip(1).enumerate() {
+                        if escaping {
+                            escaping = false;
+                            quoted_string.push(c);
+                        } else {
+                            if c == 0x5c // backslash
+                            {
+                                escaping = true;
+                            } else if c == 0x22 // double quote
+                            {
+                                end = Some(i + 1); // cuz skipped 1 for the leading quote
+                                break;
+                            } else {
+                                quoted_string.push(c);
+                            }
+                        }
+                    }
+                    left = &left[end.ok_or(::error::ParseError::Header)? + 1..];
+                    left = split_once(left, ';').1.trim_left();
+                    // In fact, it should not be Err if the above code is correct.
+                    let quoted_string = String::from_utf8(quoted_string)
+                        .map_err(|_| ::error::ParseError::Header)?;
+                    quoted_string
                 } else {
-                    return Err(::error::ParseError::Header);
+                    // token: won't contains semicolon according to RFC 2616 Section 2.2
+                    let (token, new_left) = split_once_and_trim(left, ';');
+                    left = new_left;
+                    token.to_owned()
                 };
-
-                let val = if let Some(val) = parts.next() {
-                    val.trim()
-                } else {
+                if value.len() == 0 {
                     return Err(::error::ParseError::Header);
-                };
+                }
+                cd.parameters
+                    .push(match param_name.to_ascii_lowercase().as_str() {
```

Comment:
Can you please use `eq_ignore_ascii_case`? Know it is not so elegant but even more effectiveness 😄

Reply:
yes. But I am sure what is the idiomatic way to write that. Is a long `if` block fine?